Comments: List View

Copons opened the pull request:
Closes #13892 

This is the first code iteration of the **Comment List**, part of the **Comment Management** effort.
Both the data structure and the design are currently temporary, but this component will be helpful for the Comment Management development, while also providing a sandbox to quickly try out design changes and improvements.

The Comment List already features a bare minimum collection of state methods and Redux-ready actions.

The Comment List is currently testable at: `/comments/{siteSlug}`.

![may-23-2017 19-34-27](https://cloud.githubusercontent.com/assets/2070010/26367447/e5339274-3fee-11e7-8e8a-c18b53dea1b5.gif)

dmsnell commented on client/my-sites/comments/main.jsx:
```diff
@@ -10,21 +10,33 @@ import { localize } from 'i18n-calypso';
-				<div>Hello World!</div>
+				<div className="comments__primary">
+					<CommentNavigation siteSlug={ siteSlug } status={ status } />
+					<CommentList />
```
probably good enough to let this not be a separate component, instead to map over something like `this.props.comments`

actually I'd really like to see this be a list of comment ids instead of comments too (for loading and performance reasons) such that it looks like this…

```js
<CommentNavigation { ...{ siteSlug, status } } />
{ this.props.comments.map( ( { siteId, commentId } ) => (
	<CommentDetail { ...{ siteId, commentId } } />
) ) }
```

Copons replied:
Thinking about it, having both `CommentNavigation` and the actual list of comment sounds the most sensible choice.
At the same time, though, I'm afraid that eventually the list part might end up needing some methods of its own, and we'd have to move it into its own component anyway.

Also: yes, IDs would be great! I'll see what I can do with those!

dmsnell replied:
> I'm afraid that eventually the list part might end up needing some methods of its own

let's let that be a problem when it becomes a problem. for now, our practice is to eschew hiding render details, especially when it's as simple as it is here. we had a time early on in Calypso when we had a bunch of `renderDooHicky()` methods and they made components hard to read. we stopped doing that unless things got complicated.

gwwar replied:
Were we planning on using InfiniteList here, or using a different paging mechanism? The former can be a bit of a pain to setup.

dmsnell replied:
for the moment I'd like to explore _not_ using an infinite list because of the problems

gwwar replied:
Sure, but if that's the case I'd like us to explore some paging options sooner rather than later, since folks will have a lot of comments to moderate on an active site.

Do we go with simple paging, or try our hand at another InfiniteList implementation? The latter is often not trivial to implement, though having static item heights will help a lot.

For example if we have 100 comments, do we additively add them to the list and render them all, or do we use some more complex logic and only render what's in view, say a person looking at comments 50-80? Also, if we go with something more complex, how do we handle batch actions?

dmsnell replied:
@gwwar this is the idea behind loading the comment ids separately from the comment content. it should let us do simple paging with `commentIds.slice( page * size, ( page + 1 ) * size ).map( … )`

we implemented an infinite list in Simplenote and it didn't cause too much trouble. do you anticipate specific issues here?

gwwar replied:
Nothing in particular, but it does feel like we're reinventing the wheel a bit. I remember evaluating a few more InfiniteList implementations while doing a performance pass, but having to end up deferring to our current component. I remember liking https://github.com/orgsync/react-list

If we feel that our current InfiniteList is not up to snuff, I would recommend adding a task to  https://github.com/Automattic/wp-calypso/projects/32 along with some reasoning in the issue of why we need another implementation.

dmsnell replied:
well, shouldn't we justify why we need an infinite list before we toss it in there?

seems like we could quickly get carried away with it here considering how many comments there might be.

one thing I worry about with an infinite list here is that we introduce jack when we might have no reason to. the _number_ and _id_ of each comment is a simple and fast database query which could pull in the full list of comments (and even build the tree if we wanted). _however_, getting any amount of comment content in really starts bogging us down performance-wise. if we can load the entire "empty" list of comments up front with a quick query and fill in as we go I think we could have a much better experience

gwwar commented on client/my-sites/comments/comment-navigation/index.jsx:
```diff
@@ -0,0 +1,81 @@
+		: `/comments/${ this.props.siteSlug }`;
+
+	isSelectedStatus = status => DEFAULT_STATUS === status
+		? ! this.props.status || status === this.props.status
```
I think we can avoid the ternary here by updating the default status to point at `pending`
```jsx
status defaultProps = {
   status: 'pending',
}
```

gwwar commented on client/my-sites/comments/comment-navigation/index.jsx:
```diff
@@ -0,0 +1,81 @@
+						{ translate( 'Pending' ) }
+					</NavItem>
+					<NavItem
+						count={ 15 }
```
As, I recall we were going to punt on counts for m1, but let me know otherwise.

gwwar commented:
This is feeling pretty good so far. I did notice that i could select text by double clicking on one of the labels:
![screen shot 2017-05-22 at 3 59 30 pm](https://cloud.githubusercontent.com/assets/1270189/26331859/0847bb38-3f08-11e7-888b-9da626972540.png)

In any case, let's keep #13892 open, but feel free to land this when you're happy with what's here since this is backed by a feature flag. Also, let us know if there are any other UI pieces that can be worked on in parallel.

dmsnell commented on client/blocks/comment-detail/docs/mock-data.js:
```diff
@@ -0,0 +1,35 @@
+export const mockComment = {
+	content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
+	date: '2017-05-12 16:00:00',
+	id: 12345678,
```
might be helpful to match this up with at least one in the list above

Copons commented:
@gwwar @dmsnell I've updated as per your requests, and added a rough Bulk Edit view, which was the one thing that was missing.
It doesn't really have much logic underneath, nor styling (see GIF), but it should be enough to work on it.

gwwar commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -26,16 +28,21 @@ export class CommentDetail extends Component {
 		commentContent: PropTypes.string,
 		commentDate: PropTypes.string,
-		commentId: PropTypes.number,
+		commentId: PropTypes.number.isRequired,
```
We may need to relax isRequired for `commentId` and `siteId` when we visit this route on a clean cache.

gwwar commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -172,4 +188,32 @@ export class CommentDetail extends Component {
+	const comment = mockComment;
+
+	return {
+		authorAvatarUrl: comment.author.avatarUrl,
```
In general we should guard against assuming that an object is well formed. For example let's say comment is `null` on an empty cache.

For more than one level, something like would do the trick.
```jsx
get( comment, 'author.avatarUrl' );
```

We may want to create new selectors if we end up using these properties in multiple places.

gwwar commented:
@Copons Great work. Feel free to land if you're ready.

This looks like a great starting point 👍
